planetiler-core/src/main/java/com/onthegomap/planetiler/pmtiles/WriteablePmtiles.java

```diff
@@ -0,0 +1,616 @@
+        return temp;
+      }
+      leafSize *= 1.2;
+    }
+  }
+
+  private WriteablePmtiles(SeekableByteChannel channel) throws IOException {
+    this.out = channel;
+    out.write(ByteBuffer.wrap(new byte[INIT_SECTION]));
+  }
+
+  public static WriteablePmtiles newWriteToFile(Path path) throws IOException {
+    return new WriteablePmtiles(
+      FileChannel.open(path, StandardOpenOption.CREATE, StandardOpenOption.READ, StandardOpenOption.WRITE));
+  }
+
+  public static WriteablePmtiles newWriteToMemory(SeekableInMemoryByteChannel bytes) throws IOException {
+    return new WriteablePmtiles(bytes);
+  }
+
+  @Override
+  public TileOrder tileOrder() {
+    return TileOrder.HILBERT;
+  }
+
+  @Override
+  public void initialize(PlanetilerConfig config, TileArchiveMetadata tileArchiveMetadata, LayerStats layerStats) {
+    this.layerStats = layerStats;
+    this.tileArchiveMetadata = tileArchiveMetadata;
+  }
+
+  @Override
+  public void finish(PlanetilerConfig config) {
+    if (!isClustered) {
+      Collections.sort(entries);
+    }
+    try {
+      Directories archiveDirs = makeRootLeaves(entries);
+      byte[] jsonBytes = new JsonMetadata(layerStats.getTileStats(), tileArchiveMetadata.getAll()).toBytes();
+      jsonBytes = Gzip.gzip(jsonBytes);
+
+      Envelope envelope = config.bounds().latLon();
+
+      Header header = new Header(
+        (byte) 3,
+        HEADER_LEN,
+        archiveDirs.root.length,
+        INIT_SECTION + currentOffset,
+        jsonBytes.length,
+        INIT_SECTION + currentOffset + jsonBytes.length,
+        archiveDirs.leaves.length,
+        INIT_SECTION,
+        currentOffset,
+        numAddressedTiles,
+        entries.size(),
+        hashToOffset.size() + numUnhashedTiles,
+        isClustered,
+        Compression.GZIP,
+        Compression.GZIP,
+        TileType.MVT,
+        (byte) config.minzoom(),
+        (byte) config.maxzoom(),
+        (int) (envelope.getMinX() * 10_000_000),
+        (int) (envelope.getMinY() * 10_000_000),
+        (int) (envelope.getMaxX() * 10_000_000),
+        (int) (envelope.getMaxY() * 10_000_000),
+        (byte) Math.ceil(GeoUtils.getZoomFromLonLatBounds(envelope)),
+        (int) ((envelope.getMinX() + envelope.getMaxX()) / 2 * 10_000_000),
+        (int) ((envelope.getMinY() + envelope.getMaxY()) / 2 * 10_000_000)
+      );
+
+      out.write(ByteBuffer.wrap(jsonBytes));
+      out.write(ByteBuffer.wrap(archiveDirs.leaves));
+
+      out.position(0);
+      out.write(ByteBuffer.wrap(header.toBytes()));
+      out.write(ByteBuffer.wrap(archiveDirs.root));
+
+      Format format = Format.defaultInstance();
+
+      LOGGER.info("# addressed tiles: {}", numAddressedTiles);
+      LOGGER.info("# of tile entries: {}", entries.size());
+      LOGGER.info("# of tile contents: {}", (hashToOffset.size() + numUnhashedTiles));
+      LOGGER.info("Root directory: {}B", format.storage(archiveDirs.root.length, false));
```

Comment:
Not sure how to make sonar happy... it wants `format.storage` to not be evaluated if the log level doesn't reach `info`